OrbitLinuxTracing/ThreadStateVisitor.cpp

```diff
@@ -0,0 +1,272 @@
+
+// Note: Since we use PerfEventProcessor to process perf_event_open events in order, OnNewTask,
+// OnSchedWakeup, OnContextSwitchIn, OnContextSwitchOut are expected to be called in order of
+// timestamp. But the initial thread states are retrieved (and OnInitialState is called) after the
+// perf_event_open file descriptors have been enabled, so that we don't lose thread states between
+// retrieving the initial states and enabling the file descriptors. It is then common for some of
+// the first tracepoint events to have a timestamp lower than the timestamp of initial retrieval. In
+// all these cases, we discard the previous known state (the one retrieved at the beginning, with a
+// larger timestamp) and replace it with the thread state carried by the tracepoint.
+
+void ThreadStateManager::OnInitialState(uint64_t timestamp_ns, pid_t tid,
+                                        ThreadStateSlice::ThreadState state) {
+  CHECK(!tid_open_states_.contains(tid));
+  tid_open_states_.emplace(tid, OpenState{state, timestamp_ns});
+}
+
+void ThreadStateManager::OnNewTask(uint64_t timestamp_ns, pid_t tid) {
+  constexpr ThreadStateSlice::ThreadState kNewState = ThreadStateSlice::kRunnable;
+
+  if (auto open_state_it = tid_open_states_.find(tid);
+      open_state_it != tid_open_states_.end() &&
+      timestamp_ns >= open_state_it->second.begin_timestamp_ns) {
+    ERROR("Processed task:task_newtask but thread %d was already known", tid);
+    return;
+  }
+  tid_open_states_.insert_or_assign(tid, OpenState{kNewState, timestamp_ns});
+}
+
+std::optional<ThreadStateSlice> ThreadStateManager::OnSchedWakeup(uint64_t timestamp_ns,
+                                                                  pid_t tid) {
+  constexpr ThreadStateSlice::ThreadState kNewState = ThreadStateSlice::kRunnable;
+
+  auto open_state_it = tid_open_states_.find(tid);
+  if (open_state_it == tid_open_states_.end()) {
+    ERROR("Processed sched:sched_wakeup but previous state of thread %d is unknown", tid);
+    tid_open_states_.insert_or_assign(tid, OpenState{kNewState, timestamp_ns});
+    return std::nullopt;
+  }
+
+  const OpenState& open_state = open_state_it->second;
+  if (timestamp_ns < open_state.begin_timestamp_ns) {
+    // As noted above, overwrite the thread state retrieved at the beginning.
+    tid_open_states_.insert_or_assign(tid, OpenState{kNewState, timestamp_ns});
+    return std::nullopt;
+  }
+
+  if (open_state.state == kNewState || open_state.state == ThreadStateSlice::kRunning) {
+    // It seems to be somewhat common for a thread to receive a wakeup
+    // while already in runnable or running state: disregard the state change.
+    return std::nullopt;
+  }
+
+  if (open_state.state == ThreadStateSlice::kZombie ||
+      open_state.state == ThreadStateSlice::kDead) {
+    ERROR("Processed sched:sched_wakeup for thread %d but unexpected previous state %s", tid,
+          ThreadStateSlice::ThreadState_Name(open_state.state));
```

Comment:
Hmm, I don't know. If we think that this should not happen, then we should fail. Tbh, we will never notice that there was an error, when just logging it.
If you say, that this might be a strange kernel thing, that could happen, then drop the error (or use a LOG).

Reply:
Logging it is very useful, as I monitor the service's logs when developing and testing, and from logs like these we can learn what to expect from the system. And as it's an unexpected situation, for me it's an error, which is exactly saying "it's unexpected, but we can/must handle it gracefully".

Reply:
But thats for development. Logging in production is still useful, if something goes wrong, you could check the log. But here, there will be nothing going wrong. As said in the other comment. for the moment, fine for me (but only as the service crash on CHECK, is not really visible for the user). Longer term solution, must be, to be able to monitor those errors, or have a good handling of a crash in the service, s.t. we can do checks here.

Reply:
I don't think that crashing the service just to get a report for a case that we can easily handle gracefully and that doesn't really depend on us is a good idea, even if the service crashing was handled super nicely. Of course it would be nice to get those logs from production, I totally agree on that, but that's a completely different story and I assume will be part of the work on metrics.

Reply:
Yeah mostly agree. 
But, I think it can be argued whether "that doesn't really depend on us". Those cases in the service in general are often certain assumptions on the kernel to behave that way. If it fails, it means our assumptions were wrong and we should investigate why.

But nevermind, lets keep it as it is for now. At one point in time, we need to reargue about all the ERRORS in service anyways.

Reply:
Adding a comment here because I think this is important: I'm fully with dpallotti@ on this. We should not potentially make Orbit unusable for our users (what if the way the game uses the kernel triggers a bad assumption every time?) in case that an assumption we make is wrong, in particular when we can handle that case gracefully. Logging an error for sure. CHECKs should be reserved for assumptions about our own code and in cases where not asserting may lead to worse.